src/fc/qemu/hazmat/libceph.py

```diff
@@ -0,0 +1,187 @@
+"""
+A reimplementation of the Ceph librados/librbd bindings to avoid having to
+compile against them.
+
+This helps us to be more version-neutral.
+
+"""
+
+import errno
+import json
+import shlex
+
+from fc.qemu.util import cmd
+
+
+class ImageNotFound(Exception):
+    pass
+
+
+class ImageBusy(Exception):
+    pass
+
+
+class ImageExists(Exception):
+    pass
+
+
+class Rados:
+    def __init__(self, conffile, name, log):
+        self.conffile = conffile
+        self.name = name
+        self.log = log.bind(subsystem="libceph")
+        self._ioctx = {}
+
+    def connect(self):
+        pass
+
+    def open_ioctx(self, pool):
+        if pool not in self._ioctx:
+            self._ioctx[pool] = Ioctx(self, pool)
+        return self._ioctx[pool]
+
+    def shutdown(self):
+        pass
+
+    def _ceph(self, *args, use_json=True):
+        shargs = shlex.join(args)
+        format_arg = "--format json" if use_json else ""
+        result = cmd(
+            f"ceph -c {self.conffile} --name {self.name} {format_arg} {shargs}",
+            log=self.log,
+        )
+        if use_json:
+            result = json.loads(result)
+        return result
+
+    def _rbd(self, *args, use_json=True):
+        shargs = shlex.join(args)
+        format_arg = "--format json" if use_json else ""
+        result = cmd(
+            f"rbd -c {self.conffile} --name {self.name} {format_arg} {shargs}",
+            log=self.log,
+        )
+        if use_json:
+            result = json.loads(result)
+        return result
+
+    def list_pools(self):
+        pools = self._ceph("osd", "lspools")
+        return [p["poolname"] for p in pools]
+
+
+class Ioctx:
+    """Access to a pool."""
+
+    def __init__(self, rados, name: str):
+        # the rados implementation takes the name as a str, but later returns
+        # that attribute as bytes
```

Comment:
Okay, but what are the implications of that? We apparently decide not to follow that behaviour here?

edit: Found the encoding changes in the calling code. Fine, but the decision of doing so is not self-evident from this code comment (to me at least).